drivers/counter/counter_dw_timer.c

```diff
@@ -0,0 +1,413 @@
+/*
+ * Copyright (c) 2023 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT snps_dw_timers
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/counter.h>
+#include <zephyr/spinlock.h>
+#include <zephyr/logging/log.h>
+
+#define COUNTER_DW_RESET_SUPPORT \
+	DT_ANY_INST_HAS_PROP_STATUS_OKAY(resets)
+
+#if COUNTER_DW_RESET_SUPPORT
+#include <zephyr/drivers/reset.h>
+#endif
+
+#define COUNTER_DW_CLK_MANAGER_SUPPORT \
+	DT_ANY_INST_HAS_PROP_STATUS_OKAY(clocks)
+
+#if COUNTER_DW_CLK_MANAGER_SUPPORT
+#include <zephyr/drivers/clock_control.h>
+#endif
+
+LOG_MODULE_REGISTER(dw_timer, CONFIG_COUNTER_LOG_LEVEL);
+
+static int counter_dw_timer_get_value(const struct device *timer_dev, uint32_t *ticks);
+
+/* DW APB timer register offsets */
+#define LOADCOUNT_OFST           0x0
+#define CURRENTVAL_OFST          0x4
+#define CONTROLREG_OFST          0x8
+#define EOI_OFST                 0xc
+#define INTSTAT_OFST             0x10
+
+/* free running mode value */
+#define FREE_RUNNING_MODE_VAL 0xFFFFFFFFUL
+
+/* DW APB timer control flags */
+#define TIMER_CONTROL_ENABLE_BIT    0
+#define TIMER_MODE_BIT              1
+#define TIMER_INTR_MASK_BIT         2
+
+/* DW APB timer modes */
+#define USER_DEFINED_MODE       1
+#define FREE_RUNNING_MODE       0
+
+#define DEV_CFG(_dev) ((const struct counter_dw_timer_config *)(_dev)->config)
+#define DEV_DATA(_dev) ((struct counter_dw_timer_drv_data *const)(_dev)->data)
+
```

Comment:
There is one conflict requirement  of **struct counter_config_info**_(from counter.h)_ info and **DEVICE_MMIO_ROM**_(from device_mmio.h)_ to be a first member of a driver config structure. So we decided to use **DEVICE_MMIO_NAMED_*** macros instead of DEVICE_MMIO_* macros

As we have to use the DEVICE_MMIO_NAMED* macros, device_mmio.h internally expects DEV_CFG and DEV_DATA to be defined by driver so that device_mmio.h can access the proper address by typecasting